src/target/riscv/riscv.c

```diff
@@ -5207,188 +5221,154 @@ int riscv_enumerate_triggers(struct target *target)
+	[GDB_REGNO_A4] = "a4",
+	[GDB_REGNO_A5] = "a5",
+	[GDB_REGNO_A6] = "a6",
+	[GDB_REGNO_A7] = "a7",
+	[GDB_REGNO_S2] = "s2",
+	[GDB_REGNO_S3] = "s3",
+	[GDB_REGNO_S4] = "s4",
+	[GDB_REGNO_S5] = "s5",
+	[GDB_REGNO_S6] = "s6",
+	[GDB_REGNO_S7] = "s7",
+	[GDB_REGNO_S8] = "s8",
+	[GDB_REGNO_S9] = "s9",
+	[GDB_REGNO_S10] = "s10",
+	[GDB_REGNO_S11] = "s11",
+	[GDB_REGNO_T3] = "t3",
+	[GDB_REGNO_T4] = "t4",
+	[GDB_REGNO_T5] = "t5",
+	[GDB_REGNO_T6] = "t6",
+	[GDB_REGNO_PC] = "pc",
+	[GDB_REGNO_CSR0] = "csr0",
+	[GDB_REGNO_PRIV] = "priv",
+	[GDB_REGNO_FT0] = "ft0",
+	[GDB_REGNO_FT1] = "ft1",
+	[GDB_REGNO_FT2] = "ft2",
+	[GDB_REGNO_FT3] = "ft3",
+	[GDB_REGNO_FT4] = "ft4",
+	[GDB_REGNO_FT5] = "ft5",
+	[GDB_REGNO_FT6] = "ft6",
+	[GDB_REGNO_FT7] = "ft7",
+	[GDB_REGNO_FS0] = "fs0",
+	[GDB_REGNO_FS1] = "fs1",
+	[GDB_REGNO_FA0] = "fa0",
+	[GDB_REGNO_FA1] = "fa1",
+	[GDB_REGNO_FA2] = "fa2",
+	[GDB_REGNO_FA3] = "fa3",
+	[GDB_REGNO_FA4] = "fa4",
+	[GDB_REGNO_FA5] = "fa5",
+	[GDB_REGNO_FA6] = "fa6",
+	[GDB_REGNO_FA7] = "fa7",
+	[GDB_REGNO_FS2] = "fs2",
+	[GDB_REGNO_FS3] = "fs3",
+	[GDB_REGNO_FS4] = "fs4",
+	[GDB_REGNO_FS5] = "fs5",
+	[GDB_REGNO_FS6] = "fs6",
+	[GDB_REGNO_FS7] = "fs7",
+	[GDB_REGNO_FS8] = "fs8",
+	[GDB_REGNO_FS9] = "fs9",
+	[GDB_REGNO_FS10] = "fs10",
+	[GDB_REGNO_FS11] = "fs11",
+	[GDB_REGNO_FT8] = "ft8",
+	[GDB_REGNO_FT9] = "ft9",
+	[GDB_REGNO_FT10] = "ft10",
+	[GDB_REGNO_FT11] = "ft11",
+
+	#define DECLARE_CSR(csr_name, number)[(number) + GDB_REGNO_CSR0] = #csr_name,
+	#include "encoding.h"
+	#undef DECLARE_CSR
+};
+
+static void free_reg_names(struct target *target)
+{
+	RISCV_INFO(info);
+
+	for (unsigned int i = 0; i < GDB_REGNO_COUNT; ++i) {
+		if (i > ARRAY_SIZE(default_reg_names)) {
+			free((void *)info->reg_names[i]);
+		}
+		else{
+			if (default_reg_names[i] == NULL)
+				free((void *)info->reg_names[i]);
+		}
+	}
+	free(info->reg_names);
+	info->reg_names = NULL;
+
+	free_custom_reg_names(target);
+}
+
+const char *gdb_regno_name(struct target *target, unsigned int regno)
+{
+	RISCV_INFO(info);
+
+	if (regno >= GDB_REGNO_COUNT) {
+		assert(info->custom_register_names.reg_names);
+		assert(regno - GDB_REGNO_COUNT <= info->custom_register_names.num_entries);
+		return info->custom_register_names.reg_names[regno - GDB_REGNO_COUNT];
+	}
+
+	if (!info->reg_names) {
+		info->reg_names = calloc(GDB_REGNO_COUNT, sizeof(char *));
+		memcpy(info->reg_names, default_reg_names, sizeof(default_reg_names));
+	}
+		
+	if (info->reg_names[regno])
+		return info->reg_names[regno];
+	if (regno <= GDB_REGNO_XPR31)
+		return info->reg_names[regno] = initialize_reg_name_with_prefixed_name("x", regno - GDB_REGNO_ZERO);
+	if (regno <= GDB_REGNO_V31 && regno >= GDB_REGNO_V0)
+		return info->reg_names[regno] = initialize_reg_name_with_prefixed_name("v", regno - GDB_REGNO_V0);
+	if (regno >= GDB_REGNO_CSR0 && regno <= GDB_REGNO_CSR4095)
+		return info->reg_names[regno] = initialize_reg_name_with_prefixed_name("csr", regno - GDB_REGNO_CSR0);
```

Comment:
If the user exposes its own named CSR via command `riscv expose_csrs <number>=<name>`, it seems to me that the user-provided name will not be reflected in the reg_names table. Or did I just miss that logic somewhere?

Reply:
The name of the registers entered by the user sets above (lines 5348-5352)
```
        if (regno >= GDB_REGNO_COUNT) {
                assert(info->custom_register_names.reg_names);
                assert(regno - GDB_REGNO_COUNT <= info->custom_register_names.num_entries);
                return info->custom_register_names.reg_names[regno - GDB_REGNO_COUNT];
        }
```

Reply:
Corrected.